program/src/processor/slash.rs

```diff
@@ -147,64 +150,100 @@ pub fn process_slash(
     );
 
     // Update the stake amount on both stake and config accounts.
+    //
+    // The amount slashed is taken from the active stake first; in case
+    // the amount is not enough, the remaining is then taken from the inactive
+    // stake; if the inactive stake is not enough, the remaining is ignored
+    // and the stake account is left with 0 amount.
 
     require!(
-        stake.amount >= amount,
-        StakeError::InsufficientStakeAmount,
-        "stake"
+        amount > 0,
+        StakeError::InvalidAmount,
+        "amount must be greater than 0"
     );
 
+    let mut slash_amount = min(amount, stake.amount);
+
     stake.amount = stake
         .amount
-        .checked_sub(amount)
+        .checked_sub(slash_amount)
         .ok_or(ProgramError::ArithmeticOverflow)?;
 
+    let remaining = amount.saturating_sub(slash_amount);
+
+    if remaining > 0 && stake.inactive_amount > 0 {
```

Comment:
This confused me a bit at first. Can we do a little refactoring to make this a bit easier to follow?

```rust
let mut remaining_slash_amount = amount;

// First slash active stake.
let active_stake_to_slash = min(stake.amount, remaining_slash_amount);
stake.amount = stake
    .amount
    .checked_sub(active_stake_to_slash)
    .ok_or(ProgramError::ArithmeticOverflow)?;
remaining_slash_amount = remaining_slash_amount.saturating_sub(active_stake_to_slash);

// Then, if the slash amount has not been fulfilled, slash inactive stake.
if remaining_slash_amount > 0 {
    let inactive_stake_to_slash = min(stake.inactive_amount, remaining_slash_amount);
    stake.amount = stake
        .inactive_amount
        .checked_sub(inactive_stake_to_slash)
        .ok_or(ProgramError::ArithmeticOverflow)?;
    remaining_slash_amount = remaining_slash_amount.saturating_sub(inactive_stake_to_slash);
}

// Then, if the slash amount has not been fulfilled, slash deactivating stake.
// ...
```

It's pretty similar to what you have, just a bit more clear.

Reply:
Nice, looks better now.